include/revng/ABI/FunctionType.h

```diff
@@ -0,0 +1,94 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "revng/Model/Types.h"
+
+namespace abi::FunctionType {
+
+/// Best effort `CABIFunctionType` to `RawFunctionType` conversion.
+///
+/// If `ABI` is not specified, `TheBinaryToAddSupportTypesTo.DefaultABI`
+/// is used instead.
+std::optional<model::CABIFunctionType>
+tryConvertToCABI(const model::RawFunctionType &Function,
+                 model::Binary &TheBinaryToAddSupportTypesTo,
+                 std::optional<model::ABI::Values> ABI = std::nullopt);
+
+/// Best effort `RawFunctionType` to `CABIFunctionType` conversion.
+///
+/// \note: this convertion is loosy since there's no way to represent some types
+/// in `RawFunctionType` in a reversible manner.
+std::optional<model::RawFunctionType>
+tryConvertToRaw(const model::CABIFunctionType &Function,
+                model::Binary &TheBinaryToAddSupportTypesTo);
+
+struct Layout;
+
+/// Extracts the information about argument and return value location layout
+/// from the \param Function.
+Layout layout(const model::RawFunctionType &Function);
+
+/// Extracts the information about argument and return value location layout
+/// from the \param Function.
+Layout layout(const model::CABIFunctionType &Function);
+
+/// Indicates the layout of arguments and return values of a function.
+///
+/// \note Such a layout is immutable.
+struct Layout {
+public:
+  struct ReturnValueRegisters {
+    llvm::SmallVector<model::Register::Values, 2> Registers;
+  };
+
+  struct Argument : public ReturnValueRegisters {
+  public:
+    struct StackSpan {
+      size_t Offset;
+      size_t Size;
+    };
+
+  public:
+    std::optional<StackSpan> Stack;
+  };
+
+private:
+  friend Layout layout(const model::RawFunctionType &Function);
+  friend Layout layout(const model::CABIFunctionType &Function);
+
+  Layout(llvm::SmallVector<Argument, 4> &&Arguments,
+         ReturnValueRegisters &&ReturnValue,
+         llvm::SmallVector<model::Register::Values, 24> &&CalleeSavedRegisters,
+         uint64_t FinalStackOffset) :
+    Arguments(std::move(Arguments)),
+    ReturnValue(std::move(ReturnValue)),
+    CalleeSavedRegisters(std::move(CalleeSavedRegisters)),
+    FinalStackOffset(FinalStackOffset) {}
+
+public:
+  bool verify() const;
+  size_t argumentRegisterCount() const;
+  size_t returnValueRegisterCount() const;
+
+public:
+  const llvm::SmallVector<Argument, 4> Arguments;
+  const ReturnValueRegisters ReturnValue;
+  const llvm::SmallVector<model::Register::Values, 24> CalleeSavedRegisters;
+  const uint64_t FinalStackOffset;
```

Comment:
Using `const` makes the object non-assignable, you sure you don't want to pass around `const` references instead?
Or maybe go for data hiding.